sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsFileSystem.java

```diff
@@ -76,6 +90,41 @@ protected void copy(List<GcsResourceId> srcResourceIds, List<GcsResourceId> dest
     options.getGcsUtil().copy(toFilenames(srcResourceIds), toFilenames(destResourceIds));
   }
 
+  /**
+   * Expands a pattern into {@link MatchResult}.
+   *
+   * <p>{@code gcsPattern} is expected to contain globs.
+   */
+  @VisibleForTesting
+  MatchResult expand(GcsPath gcsPattern) throws IOException {
+    String prefix = GcsUtil.getGlobPrefix(gcsPattern.getObject());
+    Pattern p = Pattern.compile(GcsUtil.globToRegexp(gcsPattern.getObject()));
+
+    LOG.debug("matching files in bucket {}, prefix {} against pattern {}", gcsPattern.getBucket(),
+        prefix, p.toString());
+
+    String pageToken = null;
+    List<Metadata> results = new LinkedList<>();
+    do {
+      Objects objects = options.getGcsUtil().listObjects(gcsPattern.getBucket(), prefix, pageToken);
+      if (objects.getItems() == null) {
+        break;
+      }
+
+      // Filter objects based on the regex.
+      for (StorageObject o : objects.getItems()) {
+        String name = o.getName();
+        // Skip directories, which end with a slash.
+        if (p.matcher(name).matches() && !name.endsWith("/")) {
+          LOG.debug("Matched object: {}", name);
```

Comment:
This is debug level logs which won't show up normally. If moved to the end, there will be additional cost of parsing the message from Metadata List.